### PR TITLE
YDA-5728: reduce queries for category statistics from O(n²) to O(1)

### PR DESCRIPTION
**Pull request by lwesterhof**

### CODE REVIEW

File: resources.py [hunk 1]
```diff
@@ -151,81 +151,86 @@ def api_resource_category_stats(ctx):
...
+    # Add the Yoda instance information as an extra row with category name YODA_INSTANCE_TOTAL.
+    # So the frontend can distinguish instance totals from real category totals.
     all_storage.append({'category': "YODA_INSTANCE_TOTAL",
                         'storage': {'total': misc.human_readable_size(instance_totals['total']),
                                     'research': misc.human_readable_size(instance_totals['research']),
                                     'vault': misc.human_readable_size(instance_totals['vault']),
                                     'revision': misc.human_readable_size(instance_totals['revision'])},
-                        'users': users})
+                        'users': {'internals': len(list(set(instance_totals['internals']))),
+                                  'externals': len(list(set(instance_totals['externals'])))}})
```
Comment (claravox) [inline resources.py]: I'm probably missing something, but why not make these sets from the beginning? (for any of the times we do `len(list(set(...)))`)

File: resources.py [hunk 2]
```diff
@@ -151,81 +151,85 @@ def api_resource_category_stats(ctx):
...
+            storage[category]['revision'] += revisions
+            storage[category]['total'] += total
+
+    # Retrieve groups and their members.
+    iter = list(genquery.Query(ctx,
+                ['USER_GROUP_NAME', 'USER_NAME'],
+                "USER_TYPE != 'rodsgroup'",
+                output=genquery.AS_LIST))
+
+    # Calaculate number of members per type per group.
```
Comment (claravox) [inline resources.py]:
```suggestion
    # Calculate number of members per type per group.
```

**Comment by kaur16** [comment]
LGTM!

File: resources.py [hunk 3]
```diff
@@ -151,81 +151,85 @@ def api_resource_category_stats(ctx):
...
-        cat_members[category] = list(set(members))
+        if category not in storage:
+            continue
 
-    cat_members['YODA_INSTANCE_TOTAL'] = list(set(members_total))
+        # Calculate category members and totals.
+        internals = set()
+        externals = set()
+        for group_name in get_groups_on_categories(ctx, [category]):
+            members.setdefault(group_name, {'internal': [], 'external': []})
```
Comment (claravox) [inline resources.py]: set?

File: resources.py [hunk 4]
```diff
@@ -151,81 +151,85 @@ def api_resource_category_stats(ctx):
...
 
-    def count_externals(members):
-        return len([member for member in members if not yoda_names.is_internal_user(member)])
+        # Deduplicate group members.
+        users = {'internals': len(internals), 'externals': len(externals)}
 
-    def count_internals(members):
-        return len([member for member in members if yoda_names.is_internal_user(member)])
+        # Count instance totals.
+        instance_totals['internals'].update(internals)
```
Comment (claravox) [inline resources.py:211]: all the `update`s should be `add` I think

**Review by claravox — APPROVED** [review]
LGTM!

**Pull request merged**